### ISSUE TITLE
Failed to export buffer to dma_buf: No such file or directory (2)

### ISSUE THREAD

**Issue by Rowisi** [ungoogled-software/ungoogled-chromium-archlinux#68]
I have compiled ungoogled chromium without `use_system_libopenjpeg2=true` 
Chromium runs fine but should I be worried about these?
```
[9707:9707:0427/063656.853336:ERROR:browser_switcher_service.cc(238)] XXX Init()
[9731:9731:0427/063657.019044:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.039402:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.040112:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041255:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041310:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041361:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041397:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041426:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041453:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.041482:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042815:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042879:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042899:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042912:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042925:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042952:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042968:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042981:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.042992:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043018:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043031:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043043:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043055:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043083:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043096:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043107:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043132:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043146:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043157:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043169:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043193:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.043216:ERROR:gbm_wrapper.cc(271)] Failed to export buffer to dma_buf: No such file or directory (2)
[9731:9731:0427/063657.047162:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.
```

**Comment by wchen342** [comment]
It seems to be GPU/UI related. The file used to be under ozone but I am not sure it is also the case in v81. @jstkdng Do you have any idea?

**Comment by Rowisi** [comment]
True, it seems GPU related. Here the log when I disabled hardware acc:
```
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to open i915 (search paths /usr/lib/dri)
failed to load driver: i915
MESA-LOADER: failed to open kms_swrast (search paths /usr/lib/dri)
failed to load driver: kms_swrast
MESA-LOADER: failed to open swrast (search paths /usr/lib/dri)
failed to load swrast driver
[6269:6269:0427/071304.119153:ERROR:browser_switcher_service.cc(238)] XXX Init()
```
I have Intel Graphics HD and Mesa is at latest version.
```
OpenGL vendor string: Intel
OpenGL renderer string: Mesa Intel(R) UHD Graphics 620 (WHL GT2)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 20.0.5
```

**Comment by Rowisi** [comment]
I have a deticated AMD GPU alongside my Integrated Intel GPU and It seems `Failed to export buffer to dma_buf` has been gone when I run chromium on my second GPU.

$ DRI_PRIME=1 LIBVA_DRIBER_NAME=radeonsi chromium
```
[10542:10542:0427/073252.506894:ERROR:browser_switcher_service.cc(238)] XXX Init()
[10566:10566:0427/073253.454109:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.
```

So this is most likely related to my Intel GPU. I have `intel-media-driver` installed with `LIBVA_DRIVER_NAME=iHD` set. I haven't compiled ungoogled chromium for more than a month so Idk which version caused this.

**Comment by wchen342** [comment]
That sounds like an AMD version of #48, not sure whether they are related but in that case setting `LIBVA_DRIBER_NAME` fixed the problem for me (I didn't have an Intel GPU on that machine though). I suspect video accelaration has some bugs we don't know.
For now probably try just disabling hardware accelaration as a workaround?

**Comment by Rowisi** [comment]
@wchen342  this is the log when I disable it though. Chromium still works somehow.
```
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to open i915 (search paths /usr/lib/dri)
failed to load driver: i915
MESA-LOADER: failed to open kms_swrast (search paths /usr/lib/dri)
failed to load driver: kms_swrast
MESA-LOADER: failed to open swrast (search paths /usr/lib/dri)
failed to load swrast driver
[6269:6269:0427/071304.119153:ERROR:browser_switcher_service.cc(238)] XXX Init()
```
I think that I will just use chromium without Hardware Acceleration since its safer.

**Comment by jstkdng** [comment]
If you haven't compiled ungoogled chromium for a month and the error just appeared, then it isn't a problem with ungoogled chromium.
Mesa had an upgrade when 81.0.4044.92-1 came out that broke vaapi for many users. After that it became kinda stable and works for some people.
That warning can be ignored as chromium uses gbm when it is used under wayland.

**Comment by Rowisi** [comment]
@jstkdng I haven't compiled or used ungoogled chromium for a month before compiling it yesterday.

Well, now my CPU usage is back at 70% while watching a video since vaapi is broken ¯\_(ツ)_/¯

**Comment by jstkdng** [comment]
Try adding this patches to your build @Rowisi 

vaapi-build-fix.patch
```patch
From 5b2ff215473e0526b5b24aeff4ad90d369b21c75 Mon Sep 17 00:00:00 2001
From: Julien Isorce <julien.isorce@chromium.org>
Date: Wed, 05 Feb 2020 17:59:59 +0000
Subject: [PATCH] Fix vaapi with GLX

The signature of ui's gl::GLImageGLX has changed a little bit
since "mplement GpuMemoryBuffers for EGL and GLX":
  https://chromium-review.googlesource.com/c/chromium/src/+/1984712

Bug: 1031269
Test: build with use_vaapi=true and run with --use-gl=desktop, see
Change-Id: I80b07294b9abdfa8233aaf79f7d9ec4c58117090
https: //chromium.googlesource.com/chromium/src.git/+/refs/heads/master/docs/gpu/vaapi.md#vaapi-on-linux
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2036494
Reviewed-by: Thomas Anderson <thomasanderson@chromium.org>
Reviewed-by: Miguel Casas <mcasas@chromium.org>
Commit-Queue: Julien Isorce <julien.isorce@chromium.org>
Cr-Commit-Position: refs/heads/master@{#738595}
---

diff --git a/media/gpu/vaapi/vaapi_picture_tfp.cc b/media/gpu/vaapi/vaapi_picture_tfp.cc
index 227c31b..b42620d 100644
--- a/media/gpu/vaapi/vaapi_picture_tfp.cc
+++ b/media/gpu/vaapi/vaapi_picture_tfp.cc
@@ -57,7 +57,7 @@
   if (make_context_current_cb_ && !make_context_current_cb_.Run())
     return false;

-  glx_image_ = new gl::GLImageGLX(size_, GL_RGB);
+  glx_image_ = new gl::GLImageGLX(size_, gfx::BufferFormat::BGRX_8888);
   if (!glx_image_->Initialize(x_pixmap_)) {
     // x_pixmap_ will be freed in the destructor.
     DLOG(ERROR) << "Failed creating a GLX Pixmap for TFP";
--- a/media/mojo/services/gpu_mojo_media_client.cc	2020-04-02 21:11:34.000000000 -0700
+++ b/media/mojo/services/gpu_mojo_media_client.cc	2020-04-09 00:44:58.871366432 -0700
@@ -158,6 +158,7 @@
       *d3d11_supported_configs_;

 #elif BUILDFLAG(USE_CHROMEOS_MEDIA_ACCELERATION)
+#if defined(OS_CHROMEOS)
   if (base::FeatureList::IsEnabled(kChromeosVideoDecoder)) {
     if (!cros_supported_configs_) {
       cros_supported_configs_ =
@@ -167,6 +168,7 @@
         *cros_supported_configs_;
     return supported_config_map;
   }
+#endif  //defined(OS_CHROMEOS)
 #endif

   auto& default_configs =
```
vdpau-support.patch
```patch
--- a/media/gpu/vaapi/vaapi_video_decode_accelerator.cc
+++ b/media/gpu/vaapi/vaapi_video_decode_accelerator.cc
@@ -641,6 +641,7 @@ void VaapiVideoDecodeAccelerator::AssignPictureBuffers(
   // |vpp_vaapi_wrapper_| for VaapiPicture to DownloadFromSurface() the VA's
   // internal decoded frame.
   if (buffer_allocation_mode_ != BufferAllocationMode::kNone &&
+      buffer_allocation_mode_ != BufferAllocationMode::kWrapVdpau &&
       !vpp_vaapi_wrapper_) {
     vpp_vaapi_wrapper_ = VaapiWrapper::Create(
         VaapiWrapper::kVideoProcess, VAProfileNone,
@@ -665,7 +666,8 @@ void VaapiVideoDecodeAccelerator::AssignPictureBuffers(
     PictureBuffer buffer = buffers[i];
     buffer.set_size(requested_pic_size_);
     std::unique_ptr<VaapiPicture> picture = vaapi_picture_factory_->Create(
-        (buffer_allocation_mode_ == BufferAllocationMode::kNone)
+        ((buffer_allocation_mode_ == BufferAllocationMode::kNone) ||
+         (buffer_allocation_mode_ == BufferAllocationMode::kWrapVdpau))
             ? vaapi_wrapper_
             : vpp_vaapi_wrapper_,
         make_context_current_cb_, bind_image_cb_, buffer);
@@ -1093,6 +1095,12 @@ VaapiVideoDecodeAccelerator::GetSupportedProfiles() {

 VaapiVideoDecodeAccelerator::BufferAllocationMode
 VaapiVideoDecodeAccelerator::DecideBufferAllocationMode() {
+  // NVIDIA blobs use VDPAU
+  if (VaapiWrapper::GetImplementationType() == VAImplementation::kNVIDIAVDPAU) {
+    LOG(INFO) << "VA-API driver on VDPAU backend";
+    return BufferAllocationMode::kWrapVdpau;
+  }
+
   // TODO(crbug.com/912295): Enable a better BufferAllocationMode for IMPORT
   // |output_mode_| as well.
   if (output_mode_ == VideoDecodeAccelerator::Config::OutputMode::IMPORT)
@@ -1105,7 +1113,7 @@ VaapiVideoDecodeAccelerator::DecideBufferAllocationMode() {
   // depends on the bitstream and sometimes it's not enough to cover the amount
   // of frames needed by the client pipeline (see b/133733739).
   // TODO(crbug.com/911754): Enable for VP9 Profile 2.
-  if (IsGeminiLakeOrLater() &&
+  if (false && IsGeminiLakeOrLater() &&
       (profile_ == VP9PROFILE_PROFILE0 || profile_ == VP8PROFILE_ANY)) {
     // Add one to the reference frames for the one being currently egressed, and
     // an extra allocation for both |client_| and |decoder_|, see
--- a/media/gpu/vaapi/vaapi_video_decode_accelerator.h
+++ b/media/gpu/vaapi/vaapi_video_decode_accelerator.h
@@ -204,6 +204,7 @@ class MEDIA_GPU_EXPORT VaapiVideoDecodeAccelerator
     // Using |client_|s provided PictureBuffers and as many internally
     // allocated.
     kNormal,
+    kWrapVdpau,
   };

   // Decides the concrete buffer allocation mode, depending on the hardware
--- a/media/gpu/vaapi/vaapi_wrapper.cc
+++ b/media/gpu/vaapi/vaapi_wrapper.cc
@@ -131,6 +131,9 @@ media::VAImplementation VendorStringToImplementationType(
   } else if (base::StartsWith(va_vendor_string, "Intel iHD driver",
                               base::CompareCase::SENSITIVE)) {
     return media::VAImplementation::kIntelIHD;
+  } else if (base::StartsWith(va_vendor_string, "Splitted-Desktop Systems VDPAU",
+                              base::CompareCase::SENSITIVE)) {
+    return media::VAImplementation::kNVIDIAVDPAU;
   }
   return media::VAImplementation::kOther;
 }
--- a/media/gpu/vaapi/vaapi_wrapper.h
+++ b/media/gpu/vaapi/vaapi_wrapper.h
@@ -79,6 +79,7 @@ enum class VAImplementation {
   kIntelIHD,
   kOther,
   kInvalid,
+  kNVIDIAVDPAU,
 };

 // This class handles VA-API calls and ensures proper locking of VA-API calls
```
These are the patches from chromium-vaapi. I have them in my fork and vaapi is working as expected.
This video https://www.youtube.com/watch?v=LXb3EKWsInQ plays without stuttering at 1080p60 and my cpu is at ~30%

**Comment by Rowisi** [comment]
@jstkdng I have tried to install your OBS package and VAAPI works fine, thank you.

Next time when I compile Ungoogled chromium, I'll make sure I'll add these patches.

**Comment by wchen342** [comment]
@Rowisi Glad it works for you.

@jstkdng Maybe you can add those patches here too?

**Comment by jstkdng** [comment]
@wchen342 It's not like those patches are my intellectual property or something, just take them.
I'd PR but my fork has diverged quite a bit from here.

**Comment by Rowisi** [comment]
It seems the VAAPI bug is fixed in `81.0.4044.122` build without using those patches, closing.
